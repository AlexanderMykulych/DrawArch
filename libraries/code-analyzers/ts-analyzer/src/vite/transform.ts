import { analyzeCode } from '../analysis/analyzeCode';
import { CodeInfoType } from '../analysis/types';
import type { Entity, EntityLogLocation } from '../analysis/types';


export async function transform(code: string, id: string): Promise<string> {
  const codeInfos = await analyzeCode({
    code,
    fileName: id,
  });

  const entities = codeInfos
    .filter((x): x is Entity & EntityLogLocation => x.type === CodeInfoType.Entity && !!x.logStart)
    .sort((a, b) => a.logStart > b.logStart ? -1 : 1);


  entities.forEach(entity => {
    const pos = entity.logStart + 1
    code = code.slice(0, pos) + `/* ${entity.id} */` + code.slice(pos);
  });

  return code;
}
