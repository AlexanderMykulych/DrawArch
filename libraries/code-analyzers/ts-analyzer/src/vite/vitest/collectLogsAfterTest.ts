import { ActionEntity, ActionEntityType, CodeInfo, CodeInfoType, Relationship, RelationType } from "../../analysis/types"
import { useNeo4j } from "../../neo4j/useNeo4j"
import type { LogItem, StartLogItem } from "../shim"


export async function collectLogsAfterTest(insertToDb: true): Promise<void>
export async function collectLogsAfterTest(): Promise<CodeInfo[]>
export async function collectLogsAfterTest(insertToDb?: true): Promise<CodeInfo[] | void> {
  if (!globalThis.__getLogs) {
    return
  }

  const logs = globalThis.__getLogs()

  const codeInfos = logsToCodeInfo(logs)

  if (insertToDb) {
    const db = useNeo4j()
    await db.insertCodeInfos(codeInfos);
  } else {
    return codeInfos
  }
}


export function logsToCodeInfo(logs: LogItem[]): CodeInfo[] {
  return logs
    .flatMap<CodeInfo>((item, index, items) => {
      const result = []
      if (item.type === 'end') {
        const callEntity: ActionEntity = {
          type: CodeInfoType.Entity,
          entityType: ActionEntityType.Call,
          id: getCallActionId(item, index),
          executionTime: item.execTime,
        }
        const callRelation: Relationship = {
          type: CodeInfoType.Relationship,
          relationType: RelationType.CallEntity,
          from: callEntity.id,
          to: item.id,
        }

        result.push(callEntity, callRelation)

        const [parentStartItem, parentIndex] = findParentStartLogFromIndex(items, index, item.id)

        if (parentStartItem) {
          const parentRelation: Relationship = {
            type: CodeInfoType.Relationship,
            relationType: RelationType.Parent,
            from: getCallActionId(parentStartItem, parentIndex),
            to: callEntity.id,
          }
          result.push(parentRelation)
        }
      }
      return result
    })
}

export function findStartLogFromIndex(items: LogItem[], index: number, id: string): StartLogItem {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id === id) {
      return startItem
    }
  }
  throw 'Can not find start element for: ' + id
}

export function findParentStartLogFromIndex(items: LogItem[], index: number, id: string): [StartLogItem, number] | [null, null] {
  for (let i = index - 1; i >= 0; i--) {
    const startItem = items[i]
    if (startItem.type === 'start' && startItem.id !== id) {
      return [startItem, i]
    }
  }

  return [null, null]
}

export function getCallActionId(item: LogItem, index: number): string {
  // return `call:${item.id}`
  return `call[${index}]:${item.id}`
}