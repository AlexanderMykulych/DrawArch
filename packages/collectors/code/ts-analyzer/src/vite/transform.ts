import { analyzeCode } from '../analysis/analyzeCode';
import { CodeInfoType } from '../analysis/types';
import type { Entity, EntityLogLocation } from '../analysis/types';

export type TransformConfigInjector = (args: string) => string
export type TransformConfigAfter = (code: string) => string
export type TransformConfigSkip = (id: string) => boolean

export type TransformConfig = {
  startInject?: TransformConfigInjector
  endInject?: TransformConfigInjector
  afterTransform?: TransformConfigAfter
  skipTransformation?: TransformConfigSkip
}
export async function transform(code: string, id: string, config?: TransformConfig): Promise<string | undefined> {
  if (config?.skipTransformation && config.skipTransformation(id)) {
    return
  }
  const codeInfos = await analyzeCode({
    code,
    fileName: id,
  });

  const entities = codeInfos
    .filter((x): x is Entity & EntityLogLocation =>
      x.type === CodeInfoType.Entity && !!x.logStart && !!x.logEnd
  )
  const startEntities = entities.map(x => ({
    ...x,
    injectIndex: x.logStart + 1,
    injectType: 'start',
  }))
  const endEntities = entities.map(x => ({
    ...x,
    injectIndex: x.logEnd - 1,
    injectType: 'end',
  }))
  const sortedEntities = [...startEntities, ...endEntities]
    .sort((a, b) => a.injectIndex > b.injectIndex ? -1 : 1);

  const defaultInject: TransformConfigInjector = args => `/* ${args} */`
  const startInject: TransformConfigInjector = config?.startInject
    ?? defaultInject
  
  const endInject: TransformConfigInjector = config?.endInject
    ?? defaultInject
  
  sortedEntities.forEach(entity => {
    const pos = entity.injectIndex
    const inject = entity.injectType === 'start' ? startInject : endInject

    const before = code.slice(0, pos)
    const after = code.slice(pos)

    if (entity.injectType === 'start') {
      code = `${before}
try {
${inject(JSON.stringify(entity.id))}
${after}
`
    } else {
      code = `${before}
} finally {
  ${inject(JSON.stringify(entity.id))}
}
${after}`
    }
    
  });

  if (config?.afterTransform) {
    code = config.afterTransform(code)
  }
  return code;
}
