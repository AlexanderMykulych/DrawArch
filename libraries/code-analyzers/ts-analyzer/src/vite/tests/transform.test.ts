import { test, expect } from 'vitest'
import { transform } from '../transform'

test('should transform ts code', async () => {
  const code = `
function fn1() {
  return 1
}
const obj = {
  fn1() {
    return 2
  },
  fn2: () => 3,
  fn3: () => { return 4 },
  fn4: () =>
    {
      return 4 },
}
class A {
  method() {

    return 5
  }
}
const fn5 = () => { return 6 }
  `
  const fileName = 'index.ts'

  const resultCode = await transform(code, fileName)

  expect(resultCode).toMatchSnapshot()
})
