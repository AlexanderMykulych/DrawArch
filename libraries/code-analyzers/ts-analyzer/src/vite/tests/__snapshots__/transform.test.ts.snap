// Vitest Snapshot v1

exports[`should transform ts code 1`] = `
"
function fn1() {
try {
/* \\"/index.ts/fn1\\" */

  return 1

} finally {
  /* \\"/index.ts/fn1\\" */
}
}
const obj = {
  fn1() {
try {
/* \\"/index.ts/obj/fn1\\" */

    return 2
  
} finally {
  /* \\"/index.ts/obj/fn1\\" */
}
},
  fn2: () => 3,
  fn3: () => { return 4 },
  fn4: () =>
    {
      return 4 },
}
class A {
  method() {
try {
/* \\"/index.ts/method\\" */

    return 5
  
} finally {
  /* \\"/index.ts/method\\" */
}
}
}
const fn5 = () => { return 6 }
  


"
`;
